libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -899,3 +952,435 @@ float AP_GPS::get_lag(uint8_t instance) const
         return drivers[instance]->get_lag();
     }
 }
+
+/*
+ calculate the weightings used to blend GPs location and velocity data
+*/
+void AP_GPS::calc_blend_weights(void)
+{
+    // zero the blend weights
+    memset(&_blend_weights, 0, sizeof(_blend_weights));
+
+    // determine how many GPS receivers are returning a fix and the instance of the first receiver that is
+    uint8_t first_instance = 0;
+    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+        if (state[i].status > NO_FIX) {
+            num_instances = i+1;
+            if (first_instance == 0) {
+                first_instance = i;
+            }
+        }
+    }
+
+    // If not enough receivers to do blending or blending has  been disabled, set the primary instance and weighting to the first receiver detected
+    if (num_instances < 2 || _auto_switch != 2) {
+        primary_instance = first_instance;
+        _output_is_blended = false;
+        _blend_weights[primary_instance] = 1.0f;
+        return;
+    }
+
+    // Use the oldest non-zero time, but if time difference is excessive, use newest to prevent a disconnected receiver from blocking updates
+    uint32_t max_ms = 0; // newest non-zero system time of arrival of a GPS message
+    uint32_t min_ms = -1; // oldest non-zero system time of arrival of a GPS message
+    uint8_t newest_instance = 0; // index of receiver with newest time stamp
+    int16_t max_rate_ms = 0; // largest update interval of a GPS receiver
+    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+        // Find largest and smallest times
+        if (state[i].last_gps_time_ms > max_ms) {
+            max_ms = state[i].last_gps_time_ms;
+            newest_instance = i;
+        }
+        if ((state[i].last_gps_time_ms < min_ms) && (state[i].last_gps_time_ms > 0)) {
+            min_ms = state[i].last_gps_time_ms;
+        }
+        if (_rate_ms[i] > max_rate_ms) {
+            max_rate_ms = _rate_ms[i];
```

Comment:
@priseborough So my concern on this actually is that there are drivers right now where the driver requests 10Hz updates regardless of the parameter, and no user knows that they should be setting this parameter. I guess we could have some of those drivers instead call the set parameter to override to 10Hz which might work, but this is a very weak to incorrect assumption for any driver other then u-blox. (Indeed I was actually planning to hide this parameter inside the ublox layer with the parameter hiding stuff @tridge was talking about).

Would instead tracking the update time between instances in the AP_GPS layer as an actual calculated number be fine for your use case? One of the last things with #2879 is to actually track this in the AP_GPS layer so we can publish if it's healthy or not. We could track this as a delta update time that you could look into. If that's fine I can get that sent in relatively soon.